examples/bloom-embeddings.ipynb

```diff
@@ -0,0 +1,729 @@
+   "execution_count": 1,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "executionInfo": {
+     "elapsed": 3527,
+     "status": "ok",
+     "timestamp": 1648197799884,
+     "user": {
+      "displayName": "Vincent D. Warmerdam",
+      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14Gh4KYzhhhK0YDTnAQsUIaQPw-0dKIP-kLBID7nFdQ=s64",
+      "userId": "05641618555626735638"
+     },
+     "user_tz": -60
+    },
+    "id": "ssfd1qsSxtRS",
+    "outputId": "7c6e6585-2362-4be2-da05-db00a0307fe6",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%pip install mmh3 numpy"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The Bloom embeddings algorithm\n",
+    "\n",
+    "In a normal embedding table, each word-string is mapped to a distinct ID.\n",
+    "Usually these IDs will be sequential, so if you have a vocabulary of 100 words,\n",
+    "your words will be mapped to numbers `range(100)`. The sequential IDs can then\n",
+    "adrianeboyd marked this conversation as resolved.\n",
+    "be used as indices into an embedding table: if you have 100 words in your\n",
+    "adrianeboyd marked this conversation as resolved.\n",
+    "vocabulary, you have 100 rows in the table, and each word receives its own\n",
+    "vector.\n",
+    "\n",
+    "However, there's no limit to the number of unique words that might occur in a\n",
+    "sample of text, while we definitely want a limited number of rows in our\n",
+    "embedding table. Some of the rows in our table will therefore need to be shared\n",
+    "between multiple words in our vocabulary. One obvious solution is to set aside a\n",
+    "single vector in the table. Words 0-98 will each receive their own vector, while\n",
+    "all other words are assigned to vector 99.\n",
+    "\n",
+    "However, this asks vector 99 to do a lot of work. What if we gave more vectors\n",
+    "to the unknown words?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "executionInfo": {
+     "elapsed": 10,
+     "status": "ok",
+     "timestamp": 1648197799885,
+     "user": {
+      "displayName": "Vincent D. Warmerdam",
+      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14Gh4KYzhhhK0YDTnAQsUIaQPw-0dKIP-kLBID7nFdQ=s64",
+      "userId": "05641618555626735638"
+     },
+     "user_tz": -60
+    },
+    "id": "Eb895XpR-VUB"
+   },
+   "outputs": [],
+   "source": [
+    "def get_row(word_id, number_vector=100, number_oov=10):\n",
+    "    if word_id < (number_vector - number_oov):\n",
+    "        return word_id\n",
+    "    else:\n",
+    "        return number_vector + (word_id % number_oov)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This gives the model a little more resolution for the unknown words. If all\n",
+    "out-of-vocabulary words are assigned the same vector, then they'll all look\n",
```

Comment:
In a sense I agree, but in another sense since the encoding is most of the time contextual, models learn to properly contextualize `UNK`. This is why for example a pointer mechanism is good at a copying unseen words.

Reply:
I think it's okay that this explanation is only focused on the embed part.